### ISSUE TITLE
[REVIEW]: Synch: A framework for concurrent data-structures and benchmarks

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3143]
**Submitting author:** @nkallima (<a href="http://orcid.org/0000-0002-0331-1475">Nikolaos D. Kallimanis</a>)
**Repository:** <a href="https://github.com/nkallima/sim-universal-construction" target ="_blank">https://github.com/nkallima/sim-universal-construction</a>
**Version:** v3.0.0
**Editor:** @danielskatz
**Reviewers:** @jarrah42, @williamfgc
**Archive:** <a href="https://doi.org/10.5281/zenodo.5347958" target="_blank">10.5281/zenodo.5347958</a>

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/07bf35ba1bd72c38cc8076fee6864409/status.svg)](https://joss.theoj.org/papers/07bf35ba1bd72c38cc8076fee6864409)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/07bf35ba1bd72c38cc8076fee6864409"><img src="https://joss.theoj.org/papers/07bf35ba1bd72c38cc8076fee6864409/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/07bf35ba1bd72c38cc8076fee6864409/status.svg)](https://joss.theoj.org/papers/07bf35ba1bd72c38cc8076fee6864409)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/nkallima/sim-universal-construction" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@williamfgc & @jarrah42, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @danielskatz know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/nkallima/sim-universal-construction">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@nkallima) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages? Mentioned in the background section
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/nkallima/sim-universal-construction">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@nkallima) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @hkaiser, @jarrah42 it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.18 s (634.2 files/s, 43956.5 lines/s)
--------------------------------------------------------------------------------
Language                      files          blank        comment           code
--------------------------------------------------------------------------------
C                                60            777             58           4435
C/C++ Header                     39            283             75           1155
Bourne Shell                      4             45             24            366
Markdown                          6            129              0            347
YAML                              3              6              1            149
Bourne Again Shell                1              4              3             19
make                              1              7              0             18
--------------------------------------------------------------------------------
SUM:                            114           1251            161           6489
--------------------------------------------------------------------------------


Statistical information for the repository 'c14b6e0062df02a63525de43' was
gathered on 2021/04/01.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Nick Kallimanis                 23           636            462            2.24
Nikolaos D. Kalliman            61          2295           1666            8.08
Nikolaos Kallimanis              3          5629           9803           31.48
Spyridon N. Agathos              3            27            170            0.40
nkallima                        25         10594           8774           39.51
nkallima@gmail.com               7          8751            218           18.29

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Nick Kallimanis              16            2.5          5.5                0.00
Nikolaos D. Kalliman       2077           90.5          3.5                2.79
Spyridon N. Agathos          89          329.6          4.2                1.12
nkallima                   4510           42.6         38.0                1.57
nkallima@gmail.com           91            1.0         99.6                5.49
```

**Comment by whedon** [comment]
PDF failed to compile for issue #3143 with the following error: 

 Can't find any papers to compile :-(

**Comment by danielskatz** [comment]
@whedon generate pdf from branch joss-branch

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss-branch. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03143/joss.03143/10.21105.joss.03143.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03143/joss.03143/10.21105.joss.03143.pdf) :page_facing_up: :point_left:

**Comment by danielskatz** [comment]
@whedon check references from branch joss-branch

**Comment by whedon** [comment]
```
Attempting to check references... from custom branch joss-branch
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1145/2145816.2145849 is OK
- 10.1145/1989493.1989549 is OK
- 10.4230/LIPIcs.OPODIS.2017.8 is OK
- 10.1145/3210377.3210408 is OK
- 10.1109/IPPS.1994.288305 is OK
- 10.1145/103727.103729 is OK
- 10.1145/248052.248106 is OK

MISSING DOIs

- None

INVALID DOIs

- https://doi.org/10.1007/s00224-013-9491-y is INVALID because of 'https://doi.org/' prefix
- 10.1.1.159.183 is INVALID
- https://doi.org/10.1007/978-3-642-32820-6_64 is INVALID because of 'https://doi.org/' prefix
```

**Comment by danielskatz** [comment]
@hkaiser and @jarrah42 - Thanks for agreeing to review this submission.

This is the review thread for the paper. All of our communications will happen here from now on.

Both reviewers have checklists at the top of this thread with the JOSS requirements. As you go over the submission, please check any items that you feel have been satisfied. There are also links to the JOSS reviewer guidelines.

Please read the first couple of comments in this issue carefully, so that you can accept the invitation from JOSS and be able to check items, and so that you don't get overwhelmed with notifications from other activities in JOSS.

The JOSS review is different from most other journals. Our goal is to work with the authors to help them meet our criteria instead of merely passing judgment on the submission. As such, the reviewers are encouraged to submit issues and pull requests on the software repository. When doing so, please mention openjournals/joss-reviews#3143 so that a link is created to this thread (and I can keep an eye on what is happening). Please also feel free to comment and ask questions on this thread. In my experience, it is better to post comments/questions/suggestions as you come across them instead of waiting until you've reviewed the entire package.

We aim for reviews to be completed within about 2-4 weeks. Please let me know if you require some more time. We can also use Whedon (our bot) to set automatic reminders if you know you'll be away for a known period of time.

Please feel free to ping me (@danielskatz) if you have any questions/concerns.

**Comment by danielskatz** [comment]
👋 @nkallima - note that there a couple of references in your bib file that you can fix (by removing the `https://doi.org/` part) when you get a chance - no hurry though...  You can use the same `@whedon` comments I did above to recheck and regenerate the pdf, too.

**Comment by nkallima** [comment]
@whedon generate pdf from branch joss-branch

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss-branch. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03143/joss.03143/10.21105.joss.03143.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03143/joss.03143/10.21105.joss.03143.pdf) :page_facing_up: :point_left:

**Comment by nkallima** [comment]
@whedon check references from branch joss-branch

**Comment by whedon** [comment]
```
Attempting to check references... from custom branch joss-branch
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1145/2145816.2145849 is OK
- 10.1145/1989493.1989549 is OK
- 10.1007/s00224-013-9491-y is OK
- 10.4230/LIPIcs.OPODIS.2017.8 is OK
- 10.1145/3210377.3210408 is OK
- 10.1109/IPPS.1994.288305 is OK
- 10.1145/103727.103729 is OK
- 10.1007/978-3-642-32820-6_64 is OK
- 10.1145/248052.248106 is OK

MISSING DOIs

- None

INVALID DOIs

- 10.1.1.159.183 is INVALID
```

**Comment by nkallima** [comment]
@whedon generate pdf from branch joss-branch

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss-branch. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03143/joss.03143/10.21105.joss.03143.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03143/joss.03143/10.21105.joss.03143.pdf) :page_facing_up: :point_left:

**Comment by nkallima** [comment]
@whedon check references from branch joss-branch

**Comment by whedon** [comment]
```
Attempting to check references... from custom branch joss-branch
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1145/2145816.2145849 is OK
- 10.1145/1989493.1989549 is OK
- 10.1007/s00224-013-9491-y is OK
- 10.4230/LIPIcs.OPODIS.2017.8 is OK
- 10.1145/3210377.3210408 is OK
- 10.1142/4278 is OK
- 10.1109/IPPS.1994.288305 is OK
- 10.1145/103727.103729 is OK
- 10.1007/978-3-642-32820-6_64 is OK
- 10.1145/248052.248106 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by nkallima** [comment]
@danielskatz Thank you for pointing this out, I fixed them.

**Comment by hkaiser** [comment]
@danielskatz I will be away for the next two weeks. I will try to get things done as quickly as possible but might not be able to finalize the review before I leave.

**Comment by jarrah42** [comment]
I see two other contributors on the project, but they are not listed in the paper.

**Comment by jarrah42** [comment]
The project is missing installation instructions, user/developer documentation, and API documentation.

**Comment by nkallima** [comment]
@jarrah42  

> 
> 
> I see two other contributors on the project, but they are not listed in the paper.

First of all, thank you for your feedback. Most of the content on this repository is an extension of my research on concurrent data-structures/algorithms and performance evaluation. Recently, my colleague Spiros Agathos used the software extensively and gently contributed a few patches/fixes. Moreover, Eftychia Datsika (edatsika) improved the logo of the repository. I acknowledge Spiros and Eftychia in the Acknowledgment section of the paper.

> 
> 
> The project is missing installation instructions, user/developer documentation, and API documentation.

In the Readme file, you can find information on how to build the framework, and how to execute any of the provided benchmarks. Currently, no special installer is provided, since most of the users (i.e., researchers/developers of concurrent algorithms) heavily modify the provided data-structures and benchmarks in order to develop new algorithms. Moreover, most of the users use it on large multi-processors without administrative rights and thus an installer would not be very valuable. 
In the Readme file you can find a very simple example of the API usage. However, I understand that this is far from complete. Thus, I could provide some better documentation for the basic components but I would need some time for this. Do you think that this is OK?

**Comment by jarrah42** [comment]
> > The project is missing installation instructions, user/developer documentation, and API documentation.
> 
> In the Readme file, you can find information on how to build the framework, and how to execute any of the provided benchmarks. Currently, no special installer is provided, since most of the users (i.e., researchers/developers of concurrent algorithms) heavily modify the provided data-structures and benchmarks in order to develop new algorithms. Moreover, most of the users use it on large multi-processors without administrative rights and thus an installer would not be very valuable.
> In the Readme file you can find a very simple example of the API usage. However, I understand that this is far from complete. Thus, I could provide some better documentation for the basic components but I would need some time for this. Do you think that this is OK?

I think we can work with you until the review criteria is satisfied. It's not like a paper submission where you only get on chance.

Regarding the Readme. I would like to see a little more detail on how to build the software. E.g. where do you run the `make` command? Are there other configuration steps? Most packages these days provide some kind on installation command that will at least copy the header files and libraries to a specified location. Using CMake or similar will do this for you automatically.

I would definitely like to see complete API documentation, and better documentation on how to use the library.

Thanks!

**Comment by danielskatz** [comment]
@nkallima - As a reminder, the goal of this review is to openly and collaboratively review and improve the work until it gets to the point where it can be accepted.

**Comment by whedon** [comment]
:wave: @jarrah42, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @hkaiser, please update us on how your review is going (this is an automated reminder).

**Comment by hkaiser** [comment]
> 👋 @hkaiser, please update us on how your review is going (this is an automated reminder).

I'll be back next week and will take care of the review then.

**Comment by jarrah42** [comment]
> 👋 @jarrah42, please update us on how your review is going (this is an automated reminder).

I'm waiting for the author to respond and/or make the requested changes.

**Comment by danielskatz** [comment]
Sorry for the delay - I've been on vacation

@nkallima - can you update us on your response to @jarrah42's comments/requests?

**Comment by nkallima** [comment]
Hello @danielskatz and @jarrah42,

In the v2.5.0 branch, you can check the latest updates/improvements on the repository. I will release it to the master branch whenever the desired changes/improvements are in a good shape.

Regarding @jarrah42’s comments,
>  Regarding the Readme. I would like to see a little more detail on how to build the software. E.g. where do you run the make command?

I have added more content in README.md and I have improved some of the content. On how to built the framework you can see more on https://github.com/nkallima/sim-universal-construction/tree/v2.5.0-dev#configuring-compiling-and-installing-the-framework

> Are there other configuration steps?

As a first step, there is no need to configure anything. However, in case someone wants to better explore the performance of the framework, sections https://github.com/nkallima/sim-universal-construction/tree/v2.5.0-dev#configuring-compiling-and-installing-the-framework and https://github.com/nkallima/sim-universal-construction/tree/v2.5.0-dev#performanceoptimizations are a good start.

> Most packages these days provide some kind on installation command that will at least copy the header files and libraries to a specified location. Using CMake or similar will do this for you automatically.

Now, a simple `make install` and `make uninstall` is provided for installing and uninstalling the framework. Please check  https://github.com/nkallima/sim-universal-construction/tree/v2.5.0-dev#configuring-compiling-and-installing-the-framework

> I would definitely like to see complete API documentation, and better documentation on how to use the library.

I have started adding documentation (i.e. manpages) on the API. I have already added manpages for a few of the algorithms, a bit more time is needed in order to have a more complete API documentation.

**Comment by danielskatz** [comment]
👋 @nkallima - How are your changes going?

**Comment by danielskatz** [comment]
👋 @hkaiser - How is your review coming along?

**Comment by hkaiser** [comment]
> 👋 @hkaiser - How is your review coming along?

I'm struggling currently as the library under review builds on Linux only. I will need to build an environment somewhere to proceed.

**Comment by nkallima** [comment]
> 
> 
> 👋@nkallima - How are your changes going?
Things were a bit hectic the previous week, thus no much progress. It remains to complete the API documentation. I will need some more time for this.

**Comment by danielskatz** [comment]
👋 @nkallima - Any update?

**Comment by nkallima** [comment]
Hi @danielskatz

In https://github.com/nkallima/sim-universal-construction/commits/v2.5.0-dev, you can check the latest updates/improvements.
I have already added API documentation for a lot of header files. However, there are some header files that need to be documented.

**Comment by hkaiser** [comment]
@danielskatz I have to apologize for the delay. I was trying to find the time to set up a special environment to be able to test the library under review. I failed :( 

I have to recuse myself from being a reviewer on this paper as the library can't be built on my day-today development platform (Windows).

**Comment by danielskatz** [comment]
Ok, thanks @hkaiser

**Comment by danielskatz** [comment]
@whedon remove @hkaiser as reviewer

**Comment by whedon** [comment]
OK, @hkaiser is no longer a reviewer

**Comment by danielskatz** [comment]
👋 @nkallima - do you have any more reviewer suggestions, since we need to find one more now?

@jarrah42, if you know of anyone you could suggest, that would be also be useful

**Comment by danielskatz** [comment]
👋 @cedricchevalier19, @mkaguilera, @icalciu, @jamesreinders - Would one of you be willing to review this submission for JOSS? We carry out our checklist-driven reviews of the software and paper here in GitHub issues and follow these guidelines: https://joss.readthedocs.io/en/latest/review_criteria.html

As you can see, one of our assigned reviewers was unable to review as he didn't have access to a Linux platform for this, so we now again need to find a second reviewer.

**Comment by nkallima** [comment]
> 
> 👋 @nkallima - do you have any more reviewer suggestions, since we need to find one more now?
> 
> @jarrah42, if you know of anyone you could suggest, that would be also be useful

@danielskatz

A list of potential reviewers:
- rigtorp
- pramalhe
- orcaman

**Comment by danielskatz** [comment]
👋 @rigtorp, @pramalhe, @orcaman - Would one of you be willing to review this submission for JOSS?

As you can see, one of our assigned reviewers was unable to review as he didn't have access to a Linux platform for this, so we now again need to find a second reviewer.

**Comment by jarrah42** [comment]
You could also try @williamfgc

**Comment by danielskatz** [comment]
👋 @williamfgc - Would you be willing to review this submission for JOSS?

As you can see, one of our assigned reviewers was unable to review as he didn't have access to a Linux platform for this, so we now again need to find a second reviewer.

**Comment by williamfgc** [comment]
@danielskatz Happy to do it. Thank you and @jarrah42

**Comment by danielskatz** [comment]
Thanks!!  I'll update the issue at the top for you, and add you now

**Comment by danielskatz** [comment]
@whedon add williamfgc as reviewer

**Comment by whedon** [comment]
OK, williamfgc is now a reviewer

**Comment by danielskatz** [comment]
@whedon re-invite @williamfgc as reviewer

**Comment by whedon** [comment]
The reviewer already has a pending invite.

@williamfgc please accept the invite by clicking this link: https://github.com/openjournals/joss-reviews/invitations

**Comment by danielskatz** [comment]
@williamfgc and @jarrah42 - Thanks for agreeing to review this submission.

This is the review thread for the paper. All of our communications will happen here from now on.

Both reviewers have checklists at the top of this thread with the JOSS requirements. As you go over the submission, please check any items that you feel have been satisfied. There are also links to the JOSS reviewer guidelines.

Please read the first couple of comments in this issue carefully, so that you can accept the invitation from JOSS and be able to check items, and so that you don't get overwhelmed with notifications from other activities in JOSS.

The JOSS review is different from most other journals. Our goal is to work with the authors to help them meet our criteria instead of merely passing judgment on the submission. As such, the reviewers are encouraged to submit issues and pull requests on the software repository. When doing so, please mention openjournals/joss-reviews#3143 so that a link is created to this thread (and I can keep an eye on what is happening). Please also feel free to comment and ask questions on this thread. In my experience, it is better to post comments/questions/suggestions as you come across them instead of waiting until you've reviewed the entire package.

We aim for reviews to be completed within about 2-4 weeks. Please let me know if you require some more time. We can also use Whedon (our bot) to set automatic reminders if you know you'll be away for a known period of time.

Please feel free to ping me (@danielskatz) if you have any questions/concerns. and @jarrah42 - Thanks for agreeing to review this submission.

**Comment by williamfgc** [comment]
@nkallima I just opened issues above requesting some enhancements based on my initial review. Overall, great job and thanks for putting together a piece of functional software out of your research. I still need to go through some of the software functionalities, but these are initial requests/comments/feedback mostly to make the software more user-friendly. Bonus: *great CI and tests*.

**Comment by nkallima** [comment]
@williamfgc Thank you for your constructive comments and for the time devoted to review the repository. I am going to address soon the open issues and your comments directly in the repository.

> @nkallima I just opened issues above requesting some enhancements based on my initial review. Overall, great job and thanks for putting together a piece of functional software out of your research. I still need to go through some of the software functionalities, but these are initial requests/comments/feedback mostly to make the software more user-friendly. Bonus: _great CI and tests_.

**Comment by danielskatz** [comment]
👋 @nkallima - How are these updates going?

**Comment by danielskatz** [comment]
👋 @jarrah42 - How is this review going?

**Comment by nkallima** [comment]
> wave @nkallima - How are these updates going?

@danielskatz I have started working on the changes that williamfgc has proposed. Also, I have replied to a few comments directly in the repository's issues.

**Comment by danielskatz** [comment]
👋 All, I was on vacation last week, and am now trying to catch up.
@jarrah42 & @williamfgc - can you let me know what your reviews are blocked on at this point?
@nkallima - can you let me know where you are in responding to reviewer issues?

**Comment by nkallima** [comment]
> wave All, I was on vacation last week, and am now trying to catch up.
> jarrah42 & williamfgc - can you let me know what your reviews are blocked on at this point?
> nkallima - can you let me know where you are in responding to reviewer issues?

@danielskatz 
Please check at https://github.com/nkallima/sim-universal-construction/issues/32
and at https://github.com/nkallima/sim-universal-construction/issues/31

**Comment by nkallima** [comment]
@danielskatz, @jarrah42 & @williamfgc 
Now most of the API documentation is provided under the github-pages (https://nkallima.github.io/sim-universal-construction/index.html).
Although the most important parts are documented, a small part of the API remains to be documented.

**Comment by jarrah42** [comment]
I will take a look at progress this coming week.

**Comment by williamfgc** [comment]
@danielskatz sorry for the late reply. GitHub notification emails have been going to my spam folder for some reason. I still have to go through the manuscript and few more iterations on the code.

**Comment by danielskatz** [comment]
👋 @jarrah42, @williamfgc - any updates?

**Comment by danielskatz** [comment]
👋 @nkallima - I see a couple of open issues for you - any news on them?

**Comment by williamfgc** [comment]
> @williamfgc - any updates?
@danielskatz yes, I've been working with @nkallima on addressing some of the issues I opened. Thanks @nkallima for addressing those, it's looking good.

@nkallima I have a couple of comments on the manuscript related to the JOSS review process:

Summary: 
- **Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?**:  The Summary section does a good job in explaining some of the computational aspects. Please list some of the merits of the software itself in that section to motivate why first time user should use or contribute to the software. The narrative is up to you, but as an example I'd mention something along the lines of: "Synch is a C library, thoroughly tested targeting x86_64 POSIX systems. We provide comprehensive documentation for the first-time. Developers will find a comprehensive set of tests to ensure quality and reproducibility of the results."
 
-  **State of the field: Do the authors describe how this software compares to other commonly-used packages?** Are there any other similar efforts in terms of software functionality and framework? Adding some background information would be good to understand what the unique functionalities of Synch are for those familiar with other tools. 

I hope it helps, I don't think I have any other major points from my side. I haven't been able to verify benchmarking claims as I'm not aware or familiar with standard benchmarks or points of reference in this field. Tests run fine, though. 

Minor recommendations (not necessarily required): 

- [ ] Add a statement to indicate what the benchmark example goal is [ReadMe](https://github.com/nkallima/sim-universal-construction#api---code-example-for-a-simple-benchmark). While the section describes what the bechmarks does, it's not clear if the user is supposed to look at weak or strong scalability.

- [ ] While the code base is straight-forward. Add CI and containers (e.g. docker) to enable portability and reproducibility of the results and as I means for quick sharing.

**Comment by nkallima** [comment]
@williamfgc, thanks for your comments.
> 
> 
> > @williamfgc - any updates?
> 
> @danielskatz yes, I've been working with @nkallima on addressing some of the issues I opened. Thanks @nkallima for addressing those, it's looking good.
> 
> @nkallima I have a couple of comments on the manuscript related to the JOSS review process:
> 
> Summary:
> 
>     * **Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?**:  The Summary section does a good job in explaining some of the computational aspects. Please list some of the merits of the software itself in that section to motivate why first time user should use or contribute to the software. The narrative is up to you, but as an example I'd mention something along the lines of: "Synch is a C library, thoroughly tested targeting x86_64 POSIX systems. We provide comprehensive documentation for the first-time. Developers will find a comprehensive set of tests to ensure quality and reproducibility of the results."

I will mention this. At this point, I should also note that the Synch offers benchmarks (not only correctness tests) for evaluating the provided concurrent data-structures. 

> 
>     * **State of the field: Do the authors describe how this software compares to other commonly-used packages?** Are there any other similar efforts in terms of software functionality and framework? Adding some background information would be good to understand what the unique functionalities of Synch are for those familiar with other tools.

Thanks for pointing this out. I will add some material for similar tools.

> 
> 
> I hope it helps, I don't think I have any other major points from my side. I haven't been able to verify benchmarking claims as I'm not aware or familiar with standard benchmarks or points of reference in this field. Tests run fine, though.
> 
> Minor recommendations (not necessarily required):
> 
>     * [ ]  Add a statement to indicate what the benchmark example goal is [ReadMe](https://github.com/nkallima/sim-universal-construction#api---code-example-for-a-simple-benchmark). While the section describes what the bechmarks does, it's not clear if the user is supposed to look at weak or strong scalability.

I will discuss this with more details in the Readme file.

> 
>     * [ ]  While the code base is straight-forward. Add CI and containers (e.g. docker) to enable portability and reproducibility of the results and as I means for quick sharing.

Although providing the code in a docker image is a good idea, it is a tricky task. The main problem is that some of the provided data-structures (specifically the more advanced of them) need accurate low-level machine information, e.g. NUMA characteristics of the machine, etc. Trying to run the environment through docker may report inaccurate results in terms of performance (however the validation test should run OK), or the environment may fail to correctly detect the characteristics of the system. 

@danielskatz, I think I have addressed all the comments of jarrah42.
Now, I'm working on the remaining requests of williamfgc.

**Comment by jarrah42** [comment]
I lost my window this week, but I'll take a look next week.

**Comment by danielskatz** [comment]
@thanks @jarrah42 - it would be great to get this to a publishable state soon - it's been a fairly slow process compared to many others.

**Comment by danielskatz** [comment]
👋 @jarrah42 - just a ping...

**Comment by jarrah42** [comment]
@nkallima perhaps I missed it. The documentation only mentions x86_64 but does not mention that it also depends on Linux. I tried building on Darwin but it doesn't compile, so you should list the O/S dependency also.

**Comment by nkallima** [comment]
@jarrah42 please check the Requirements section of the README file in v2.5.0-dev branch (https://github.com/nkallima/sim-universal-construction/tree/v2.5.0-dev#requirements).

**Comment by jarrah42** [comment]
Ok I see it. Isn't this review for v2.3.1 or does that need to be changed now?

**Comment by danielskatz** [comment]
We'll make sure the version number is set correctly at the end of the process, in case it changes again. This is part of the publishing step, along with creating an archive of the repo at that point

**Comment by jarrah42** [comment]
My review is complete.

**Comment by danielskatz** [comment]
Thanks @jarrah42!

**Comment by danielskatz** [comment]
@williamfgc and @nkallima - how is the resolution of the remaining 4 items on @williamfgc's checklist coming?

**Comment by williamfgc** [comment]
@danielskatz I'm just waiting on @nkallima for the revised manuscript including more background information. Thanks for leading the review effort, BTW.

**Comment by nkallima** [comment]
@williamfgc @danielskatz 

I was working towards closing the issue https://github.com/nkallima/sim-universal-construction/issues/31.
Specifically, I have added a new document that describes the expected performance of various objects/data-structures provided by the framework. I also provide some indicative performance measurements. Please check https://github.com/nkallima/sim-universal-construction/blob/v2.5.0-dev/PERFORMANCE.md.

Now, it remains to slightly modify the naming of some parts of the API as was suggested by williamfgc in order to close the issue. Next, I will update the text of the manuscript following the suggestions of williamfgc.

**Comment by nkallima** [comment]
@whedon generate pdf from branch joss-branch

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss-branch. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03143/joss.03143/10.21105.joss.03143.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03143/joss.03143/10.21105.joss.03143.pdf) :page_facing_up: :point_left:

**Comment by nkallima** [comment]
@whedon check references from branch joss-branch

**Comment by whedon** [comment]
```
Attempting to check references... from custom branch joss-branch
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1145/2145816.2145849 is OK
- 10.1145/1989493.1989549 is OK
- 10.1007/s00224-013-9491-y is OK
- 10.4230/LIPIcs.OPODIS.2017.8 is OK
- 10.1145/3210377.3210408 is OK
- 10.1142/4278 is OK
- 10.1109/IPPS.1994.288305 is OK
- 10.1145/103727.103729 is OK
- 10.1007/978-3-642-32820-6_64 is OK
- 10.1145/248052.248106 is OK
- 10.1145/2442516.2442527 is OK
- 10.1007/978-3-319-26850-7_1 is OK
- 10.1145/1810479.1810540 is OK

MISSING DOIs

- 10.1007/978-1-4419-7719-9_6 may be a valid DOI for title: BOOST C++ libraries

INVALID DOIs

- None
```

**Comment by nkallima** [comment]
@williamfgc

Thanks for your comments and feedback!

> * **Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?**:  The Summary section does a good job in explaining some of the computational aspects. Please list some of the merits of the software itself in that section to motivate why first time user should use or contribute to the software. The narrative is up to you, but as an example I'd mention something along the lines of: "Synch is a C library, thoroughly tested targeting x86_64 POSIX systems. We provide comprehensive documentation for the first-time. Developers will find a comprehensive set of tests to ensure quality and reproducibility of the results."

I have added some text addressing this.

> * **State of the field: Do the authors describe how this software compares to other commonly-used packages?** Are there any other similar efforts in terms of software functionality and framework? Adding some background information would be good to understand what the unique functionalities of Synch are for those familiar with other tools.

I have added a related work section that describes similar packages.


> * [ ]  Add a statement to indicate what the benchmark example goal is [ReadMe](https://github.com/nkallima/sim-universal-construction#api---code-example-for-a-simple-benchmark). While the section describes what the bechmarks does, it's not clear if the user is supposed to look at weak or strong scalability.

I have added some extra material for handling this.


@danielskatz, @williamfgc, @jarrah42

Following the advice of  williamfgc, I have made some modifications on the naming of the API. Since the revised API is incompatible with the older version,s the version of the code should increase to v3.0.0. All improvements on the repository that are made during the review and revised API are now available at v.3.0.0-dev branch (https://github.com/nkallima/sim-universal-construction/tree/v3.0.0-dev).


@danielskatz
All issues of the repository are closed and I have applied all the requested changes to the text so far.

**Comment by danielskatz** [comment]
👋 @williamfgc - can you check and see if the changes from @nkallima satisfy you, and if you can check off the remaining items in your review?  And if not, please let us know what else you think is needed.

**Comment by danielskatz** [comment]
👋 @williamfgc - can you check and see if the changes from @nkallima satisfy you, and if you can check off the remaining items in your review? And if not, please let us know what else you think is needed.

**Comment by williamfgc** [comment]
@danielskatz it's all good from my side. Thank you, @jarrah42 and @nkallima for the overall process. I hope the review was helpful.

**Comment by danielskatz** [comment]
Thanks @williamfgc

**Comment by danielskatz** [comment]
@nkallima - the next step will for me to proofread the paper, and then for you to:
- [ ] Make a tagged release of your software, and list the version tag of the archived version here.
- [ ] Archive the reviewed software in Zenodo or a similar service (e.g., figshare, an institutional repository)
- [ ] Check the archival deposit (e.g., in Zenodo) has the correct metadata. This includes the title (should match the paper title) and author list (make sure the list is correct and people who only made a small fix are not on it). You may also add the authors' ORCID.
- [ ] Please list the DOI of the archived version here.

I'll let you know when I'm done proof-reading.

**Comment by danielskatz** [comment]
@whedon generate pdf from branch joss-branch

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss-branch. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03143/joss.03143/10.21105.joss.03143.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03143/joss.03143/10.21105.joss.03143.pdf) :page_facing_up: :point_left:

**Comment by danielskatz** [comment]
@whedon check references from branch joss-branch

**Comment by whedon** [comment]
```
Attempting to check references... from custom branch joss-branch
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1145/2145816.2145849 is OK
- 10.1145/1989493.1989549 is OK
- 10.1007/s00224-013-9491-y is OK
- 10.4230/LIPIcs.OPODIS.2017.8 is OK
- 10.1145/3210377.3210408 is OK
- 10.1142/4278 is OK
- 10.1109/IPPS.1994.288305 is OK
- 10.1145/103727.103729 is OK
- 10.1007/978-3-642-32820-6_64 is OK
- 10.1145/248052.248106 is OK
- 10.1145/2442516.2442527 is OK
- 10.1007/978-3-319-26850-7_1 is OK
- 10.1145/1810479.1810540 is OK

MISSING DOIs

- 10.1007/978-1-4419-7719-9_6 may be a valid DOI for title: BOOST C++ libraries

INVALID DOIs

- None
```

**Comment by danielskatz** [comment]
The missing DOI above appears incorrect, since the reference in the paper is to the software, not to a paper about the software, so we can ignore this suggestion from whedon

**Comment by danielskatz** [comment]
My suggested changes from proofreading are in https://github.com/nkallima/sim-universal-construction/pull/39

**Comment by nkallima** [comment]
@danielskatz 

> My suggested changes from proofreading are in [nkallima/sim-universal-construction#39](https://github.com/nkallima/sim-universal-construction/pull/39)

Thank you for leading the review and for your suggestions! I applied your patch in `joss-branch`.

> @nkallima - the next step will for me to proofread the paper, and then for you to:
> 
>     * [ ]  Make a tagged release of your software, and list the version tag of the archived version here.
> 
>     * [ ]  Archive the reviewed software in Zenodo or a similar service (e.g., figshare, an institutional repository)
> 
>     * [ ]  Check the archival deposit (e.g., in Zenodo) has the correct metadata. This includes the title (should match the paper title) and author list (make sure the list is correct and people who only made a small fix are not on it). You may also add the authors' ORCID.
> 
>     * [ ]  Please list the DOI of the archived version here.
> 
> 
> I'll let you know when I'm done proof-reading.

- A tagged release of the software was made. Version v3.0.0
- Software DOI: 10.5281/zenodo.5347958

**Comment by danielskatz** [comment]
@whedon set 10.5281/zenodo.5347958 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.5347958" target="_blank">10.5281/zenodo.5347958</a> is the archive.

**Comment by danielskatz** [comment]
@whedon set v3.0.0 as version

**Comment by whedon** [comment]
OK. v3.0.0 is the version.

**Comment by danielskatz** [comment]
@whedon recommend-accept from branch joss-branch

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1145/2145816.2145849 is OK
- 10.1145/1989493.1989549 is OK
- 10.1007/s00224-013-9491-y is OK
- 10.4230/LIPIcs.OPODIS.2017.8 is OK
- 10.1145/3210377.3210408 is OK
- 10.1142/4278 is OK
- 10.1109/IPPS.1994.288305 is OK
- 10.1145/103727.103729 is OK
- 10.1007/978-3-642-32820-6_64 is OK
- 10.1145/248052.248106 is OK
- 10.1145/2442516.2442527 is OK
- 10.1007/978-3-319-26850-7_1 is OK
- 10.1145/1810479.1810540 is OK

MISSING DOIs

- 10.1007/978-1-4419-7719-9_6 may be a valid DOI for title: BOOST C++ libraries

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/2560

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/2560, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true from branch joss-branch 
```

**Comment by danielskatz** [comment]
@nkallima - Unfortunately, a couple of the changes I made to expose references as text (rather than in parens) missed changing the required ";"s to better things in text. Merging https://github.com/nkallima/sim-universal-construction/pull/44 should fix these errors I made.

**Comment by nkallima** [comment]
> @nkallima - Unfortunately, a couple of the changes I made to expose references as text (rather than in parens) missed changing the required ";"s to better things in text. Merging [nkallima/sim-universal-construction#44](https://github.com/nkallima/sim-universal-construction/pull/44) should fix these errors I made.

Thanks! Done.

**Comment by danielskatz** [comment]
@whedon recommend-accept from branch joss-branch

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1145/2145816.2145849 is OK
- 10.1145/1989493.1989549 is OK
- 10.1007/s00224-013-9491-y is OK
- 10.4230/LIPIcs.OPODIS.2017.8 is OK
- 10.1145/3210377.3210408 is OK
- 10.1142/4278 is OK
- 10.1109/IPPS.1994.288305 is OK
- 10.1145/103727.103729 is OK
- 10.1007/978-3-642-32820-6_64 is OK
- 10.1145/248052.248106 is OK
- 10.1145/2442516.2442527 is OK
- 10.1007/978-3-319-26850-7_1 is OK
- 10.1145/1810479.1810540 is OK

MISSING DOIs

- 10.1007/978-1-4419-7719-9_6 may be a valid DOI for title: BOOST C++ libraries

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/2561

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/2561, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true from branch joss-branch 
```

**Comment by danielskatz** [comment]
@whedon accept deposit=true from branch joss-branch

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1432771608389771270) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/2562
1. Wait a couple of minutes, then verify that the paper DOI resolves [https://doi.org/10.21105/joss.03143](https://doi.org/10.21105/joss.03143)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by nkallima** [comment]
@danielskatz Thank you for handling the review and for your comments!

@williamfgc, @jarrah42 Many thanks for your fruitful comments and for helping me to improve the software!

**Comment by danielskatz** [comment]
Congratulations to @nkallima (Nikolaos D. Kallimanis)!!

And thanks to @jarrah42 and @williamfgc for reviewing!
We couldn't do this without you!

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.03143/status.svg)](https://doi.org/10.21105/joss.03143)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.03143">
  <img src="https://joss.theoj.org/papers/10.21105/joss.03143/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.03143/status.svg
   :target: https://doi.org/10.21105/joss.03143
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.03143/status.svg)](https://doi.org/10.21105/joss.03143)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss
